### ISSUE TITLE
JC C++ API - Hang when retrieving default_domain

### ISSUE THREAD

**Issue by RealAntithesis** [ryobg/JContainers#73]
I've built an SKSE DLL linking to the JContainers64 API. The intent is to call some of the JContainers functions to directly access JC containers (JArray, JMap, JValue). So far, I've successfully compiled a static library for my JC interface (which is a VS 2017-era toolset v141 code base), linked into my DLL, which is based on VS 2019 (v142) and CommonLibSSE (I didn't really want to have two DLLs talking to each other).

I've done this and it now compiles fine. The problem is that getting the default domain always results in a hang - it doesn't crash, it just continues waiting on a 'SingleObject'. Even just using the API Example project hangs on this code in line 60 of source.cpp.

Looking in the call stack, it seems to be getting stuck in the boost basic_deadline_timer in autorelease_queue(object_registry& registry) when trying to retrieve the default_domain in domain_master.

Any tips on proceeding? Given that it's happening with the unmodified API Example too, I don't think it's an issue with my code?
Thanks.

**Comment by ryobg** [comment]
Hi,

We can't be sure that this function is not working, or whatever code (Papyrus including) touching it, would hang too? True that the API example is not well tested... In any case, I can't check on this, as I'm just moving PCs and I have nothing set up to work on this project. For the brief look I had on this, I would think about an order of initialization and usage of JC, the issue looks like there is some concurrency happening by the time it is accessed.

**Comment by RealAntithesis** [comment]
Thanks for that. I've moved the retrieval of the default domain to when it is needed (when the jc functions are being called). To do this, I'm holding the pointer to the root jc interface in a variable at the time the PostPostLoad message comes through from the JContainers dll.

I'm not sure how else to do this since the default domain is accessed via the root interface which is only made known at load time via the skse messaging interface.

Anyway, my resulting jc functions seem to be crashing with an exception - at least the jmap_getInt function I'm trying out. I'm not sure if the issue is from caching the jc root interface/default domain, whether I'm passing wrong parameter types (strings are always const char*, and objects are always SInt32?) or whether something is up with the reflection interface. 

I can't seem to debug what is happening in jcontainers (due to the way the function was set up as a function pointer via reflection presumably).

**Comment by ryobg** [comment]
Try to use actual functions after PostPostPostPostLoad - just in case. Does any of these interface pointers you work with is a nullptr or they look like something which might work? Did you try to closely follow up on how the Example API project work first? Type of declarations, ways to obtain functions, exactly these functions (JArray-something). Your first goal is to try be able to make any stable call first and see whether it is working. What works in fact?

**Comment by RealAntithesis** [comment]
Further experimentation has raised more questions as to what is going on. I've gone back to the API Example, and noted a few things :
- After changing the location of retrieving the domain master from kPostLoad to kPostLoadGame, it no longer hangs.
- JArray count works to return the size of an array.
- But for some reason it only works a couple of times when stepping over before it stops working (e.g when moving the current executing line in VS to execute the function again), after which it just returns zero.
- It always seems to fail when stepping into the function when I'm debugging to figure out what is happening. The map.find(object handle) function returns zero, which is incorrect because the object of that number was created in papyrus (and the function worked in the previous iterations of the same debug session).

I don't know why this would fail after a couple of times of stepping over it or when stepping into it? Is there some timing / threading / vm issue going on here? 

I'm continuing to look into it (there's a few more tests I want to do).

**Comment by ryobg** [comment]
Hard to say anything with such context. It might be many things, like the garbage collection kicking in for example.

**Comment by RealAntithesis** [comment]
Thanks for that, it was indeed garbage collection. I belatedly realised that time is still running even when everything is frozen during a debug session.

I'll get back with more details, but it seems that any jc function which takes a string (whether const char* or std:string) is having trouble converting it (resulting in memory access violation). From what I can figure out in the jc code, I should be able to use either string type (so long as they are both the same for the default parameter and return type).

**Comment by RealAntithesis** [comment]
While all the template code is confusing me - I'm reaching the conclusion that the wrong function in the StringConverter struct in tes_binding.h is being called. Regardless of whether I try to pass a std::string or const char* to a JC function that takes a string parameter (such as JMap keys), it always appears to select the convert2J function which expects an skse::string_ref, and hence an access violation happens when trying to access skse::string_ref.c_str(). 

Unless the expectation here is that I should be constructing a string_ref instead of using std::string or const char*?

**Comment by ryobg** [comment]
If you have traced that you are calling `convert2J`, then it definitely expects a game string: `skse::string_ref` object. What is the call stack actually?

**Comment by RealAntithesis** [comment]
Call stack below. My JMap_nextKey() function takes std::strings, but this function calls the JMap nextKey function pointer via the reflection interface (I've tried passing both std::string and const char*).
``` 	JContainers64.dll!boost::exception_detail::refcount_ptr<boost::exception_detail::error_info_container>::get() Line 224	C++	Symbols loaded.
 	JContainers64.dll!reflection::binding::StringConverter::convert2J<collections::tes_context>(const skse::string_ref & str, const collections::tes_context & __formal) Line 46	C++	Symbols loaded.
 	JContainers64.dll!reflection::binding::state_proxy<skse::string_ref (__cdecl*)(collections::tes_context &,collections::map *,char const *,char const *)>::magick<&tes_api_3::tes_map_ext::nextKey<skse::string_ref>>::non_void_ret::tes_func(collections::tes_context & state, unsigned long <params_0>, skse::string_ref <params_1>, skse::string_ref <params_2>) Line 242	C++	Symbols loaded.
>	AH_Helper_Functions.dll!AH_JCHelper_Interface::JMap_nextKey(unsigned long obj, std::string previousKey, std::string endKey) Line 173	C++	Symbols loaded.
 	AH_Helper_Functions.dll!AHJC::PrintJMap(long jObject) Line 74	C++	Symbols loaded.
 	AH_Helper_Functions.dll!AHJC::PrintJObject(long jObject) Line 151	C++	Symbols loaded.
 	AH_Helper_Functions.dll!AH_SSE_Functions::JC_Test(RE::StaticFunctionTag * __formal, unsigned int jObject) Line 340	C++	Symbols loaded.
 	AH_Helper_Functions.dll!std::invoke<void (__cdecl*&)(RE::StaticFunctionTag *,unsigned int),RE::StaticFunctionTag *,unsigned int>(void(*)(RE::StaticFunctionTag *, unsigned int) & _Obj, RE::StaticFunctionTag * && _Arg1, unsigned int && <_Args2_0>) Line 1615	C++	Non-user code. Symbols loaded.
 	AH_Helper_Functions.dll!std::_Invoker_ret<void,1>::_Call<void (__cdecl*&)(RE::StaticFunctionTag *,unsigned int),RE::StaticFunctionTag *,unsigned int>(void(*)(RE::StaticFunctionTag *, unsigned int) & _Func, RE::StaticFunctionTag * && <_Vals_0>, unsigned int && <_Vals_1>) Line 745	C++	Non-user code. Symbols loaded.
 	AH_Helper_Functions.dll!std::_Func_impl_no_alloc<void (__cdecl*)(RE::StaticFunctionTag *,unsigned int),void,RE::StaticFunctionTag *,unsigned int>::_Do_call(RE::StaticFunctionTag * && <_Args_0>, unsigned int && <_Args_1>) Line 939	C++	Non-user code. Symbols loaded.
 	AH_Helper_Functions.dll!std::_Func_class<void,RE::StaticFunctionTag *,unsigned int>::operator()(RE::StaticFunctionTag * <_Args_0>, unsigned int <_Args_1>) Line 987	C++	Non-user code. Symbols loaded.
 	AH_Helper_Functions.dll!std::invoke<std::function<void __cdecl(RE::StaticFunctionTag *,unsigned int)> const &,RE::StaticFunctionTag *,unsigned int>(const std::function<void __cdecl(RE::StaticFunctionTag *,unsigned int)> & _Obj, RE::StaticFunctionTag * && _Arg1, unsigned int && <_Args2_0>) Line 1615	C++	Non-user code. Symbols loaded.
 	AH_Helper_Functions.dll!RE::BSScript::Impl::CallbackImpl<std::function<void __cdecl(RE::StaticFunctionTag *,unsigned int)> const &,std::tuple<unsigned int>,0,RE::StaticFunctionTag *>(const std::function<void __cdecl(RE::StaticFunctionTag *,unsigned int)> & a_func, std::tuple<unsigned int> && a_tuple, std::integer_sequence<unsigned __int64,0> __formal, RE::StaticFunctionTag * && <a_args_0>) Line 26	C++	Symbols loaded.
 	AH_Helper_Functions.dll!RE::BSScript::Impl::CallBack<std::function<void __cdecl(RE::StaticFunctionTag *,unsigned int)> const &,std::tuple<unsigned int>,RE::StaticFunctionTag *>(const std::function<void __cdecl(RE::StaticFunctionTag *,unsigned int)> & a_func, std::tuple<unsigned int> && a_tuple, RE::StaticFunctionTag * && <a_args_0>) Line 36	C++	Symbols loaded.
 	AH_Helper_Functions.dll!RE::BSScript::NativeFunction<0,void __cdecl(RE::StaticFunctionTag *,unsigned int),void,RE::StaticFunctionTag *,unsigned int>::MarshallAndDispatch(RE::BSScript::Variable & a_baseValue, RE::BSScript::Internal::VirtualMachine & a_vm, unsigned int a_stackID, RE::BSScript::Variable & a_resultValue, const RE::BSScript::StackFrame & a_frame) Line 99	C++	Symbols loaded.
 	SkyrimSE.exe!00007ff60dc509e9()	Unknown	Non-user code. Cannot find or open the PDB file.
 	AH_Helper_Functions.dll!REL::invoke<enum RE::BSScript::IFunction::CallResult (__cdecl RE::BSScript::NF_util::NativeFunctionBase::*)(RE::BSTSmartPointer<RE::BSScript::Stack,RE::BSTSmartPointerIntrusiveRefCount> const &,RE::BSScript::ErrorLogger *,RE::BSScript::Internal::VirtualMachine *,bool),RE::BSScript::NF_util::NativeFunctionBase *,RE::BSTSmartPointer<RE::BSScript::Stack,RE::BSTSmartPointerIntrusiveRefCount> const &,RE::BSScript::ErrorLogger * &,RE::BSScript::Internal::VirtualMachine * &,bool &,0>(RE::BSScript::IFunction::CallResult(RE::BSScript::NF_util::NativeFunctionBase::*)(const RE::BSTSmartPointer<RE::BSScript::Stack,RE::BSTSmartPointerIntrusiveRefCount> &, RE::BSScript::ErrorLogger *, RE::BSScript::Internal::VirtualMachine *, bool) && a_func, RE::BSScript::NF_util::NativeFunctionBase * && <a_args_0>, const RE::BSTSmartPointer<RE::BSScript::Stack,RE::BSTSmartPointerIntrusiveRefCount> & <a_args_1>, RE::BSScript::ErrorLogger * & <a_args_2>, RE::BSScript::Internal::VirtualMachine * & <a_args_3>, bool & <a_args_4>) Line 220	C++	Symbols loaded.
 	AH_Helper_Functions.dll!REL::Relocation<enum RE::BSScript::IFunction::CallResult (__cdecl RE::BSScript::NF_util::NativeFunctionBase::*)(RE::BSTSmartPointer<RE::BSScript::Stack,RE::BSTSmartPointerIntrusiveRefCount> const &,RE::BSScript::ErrorLogger *,RE::BSScript::Internal::VirtualMachine *,bool)>::operator()<RE::BSScript::NF_util::NativeFunctionBase *,RE::BSTSmartPointer<RE::BSScript::Stack,RE::BSTSmartPointerIntrusiveRefCount> const &,RE::BSScript::ErrorLogger * &,RE::BSScript::Internal::VirtualMachine * &,bool &,0>(RE::BSScript::NF_util::NativeFunctionBase * && <a_args_0>, const RE::BSTSmartPointer<RE::BSScript::Stack,RE::BSTSmartPointerIntrusiveRefCount> & <a_args_1>, RE::BSScript::ErrorLogger * & <a_args_2>, RE::BSScript::Internal::VirtualMachine * & <a_args_3>, bool & <a_args_4>) Line 873	C++	Symbols loaded.
 	AH_Helper_Functions.dll!RE::BSScript::NF_util::NativeFunctionBase::Call(const RE::BSTSmartPointer<RE::BSScript::Stack,RE::BSTSmartPointerIntrusiveRefCount> & a_stack, RE::BSScript::ErrorLogger * a_logger, RE::BSScript::Internal::VirtualMachine * a_vm, bool a_arg4) Line 107	C++	Symbols loaded.
 	SkyrimSE.exe!00007ff60dc62426()	Unknown	Non-user code. Cannot find or open the PDB file.
 	SkyrimSE.exe!00007ff60dc63073()	Unknown	Non-user code. Cannot find or open the PDB file.
 	SkyrimSE.exe!00007ff60dc57343()	Unknown	Non-user code. Cannot find or open the PDB file.
 	SkyrimSE.exe!00007ff60d322a4a()	Unknown	Non-user code. Cannot find or open the PDB file.
 	SkyrimSE.exe!00007ff60d040623()	Unknown	Non-user code. Cannot find or open the PDB file.
 	SkyrimSE.exe!00007ff60d632a81()	Unknown	Non-user code. Cannot find or open the PDB file.
 	SkyrimSE.exe!00007ff60d634c48()	Unknown	Non-user code. Cannot find or open the PDB file.
 	SkyrimSE.exe!00007ff60d60d6bd()	Unknown	Non-user code. Cannot find or open the PDB file.
 	kernel32.dll!BaseThreadInitThunk()	Unknown	Non-user code. Symbols loaded.
 	ntdll.dll!RtlUserThreadStart() + 33 bytes	Unknown	Non-user code. Cannot find or open the PDB file.
```

**Comment by ryobg** [comment]
You need `skse::string_ref` for these. As this is the game string object, you may look at your whatever-library and see whether there is an alternative, or try to copy the JC implementation.

**Comment by RealAntithesis** [comment]
Ok, thanks for that - it's all (so far) now working with skse::string_ref across a range of JC functions in JMap, JArray and JValue. For future reference, I had an unresolved symbol linking error in the skse relocation class when I tried to use string_refs. This was resolved by explicitly including skse64_common.lib as a dependency in the linker (because skse64_1_5_97.lib had this as an undefined external symbol only - as shown by dumpbin).
